redis/src/sentinel.rs

```diff
@@ -0,0 +1,795 @@
+            reconnect(cached_connection, connection_info)?;
+            cmd.query(cached_connection.as_mut().unwrap())
+        } else {
+            Err(err)
+        }
+    } else {
+        result
+    }
+}
+
+// non-async methods
+impl Sentinel {
+    /// Creates a Sentinel client performing some basic
+    /// checks on the URLs that might make the operation fail.
+    pub fn build<T: IntoConnectionInfo>(params: Vec<T>) -> RedisResult<Sentinel> {
+        if params.is_empty() {
+            fail!((
+                ErrorKind::EmptySentinelList,
+                "At least one sentinel is required",
+            ))
+        }
+
+        let sentinels_connection_info = params
+            .into_iter()
+            .map(|p| p.into_connection_info())
+            .collect::<RedisResult<Vec<ConnectionInfo>>>()?;
+
+        let mut connections_cache = vec![];
+        connections_cache.resize_with(sentinels_connection_info.len(), Default::default);
+
+        let mut async_connections_cache = vec![];
+        async_connections_cache.resize_with(sentinels_connection_info.len(), Default::default);
+
+        Ok(Sentinel {
+            sentinels_connection_info,
+            connections_cache,
+            async_connections_cache,
+            replica_start_index: random_replica_index(1000000),
+        })
+    }
+
+    /// Try to execute the given command in each sentinel, returning the result of the
+    /// first one that executes without errors. If all return errors, we return the
+    /// error of the last attempt.
+    ///
+    /// For each sentinel, we first check if there is a cached connection, and if not
+    /// we attemp to connect to it (skipping that sentinel if there is an error during
+    /// the connection). Then, we attempt to execute the given command with the cached
+    /// connection. If there is an error indicating that the connection is invalid, we
+    /// reconnect and try one more time in the new connection.
+    ///
+    fn try_all_sentinels<T: FromRedisValue>(&mut self, cmd: Cmd) -> RedisResult<T> {
+        let mut last_err = None;
+        for (connection_info, cached_connection) in self
+            .sentinels_connection_info
+            .iter()
+            .zip(self.connections_cache.iter_mut())
+        {
+            match try_single_sentinel(cmd.clone(), connection_info, cached_connection) {
+                Ok(result) => {
+                    return Ok(result);
+                }
+                Err(err) => {
+                    last_err = Some(err);
+                }
+            }
+        }
+
+        // We can unwrap here because we know there is at least one connection info.
+        // Maybe we should have a more specific error for this situation?
+        Err(last_err.expect("There should be at least one connection info"))
+    }
+
+    /// Get a list of all masters (using the command SENTINEL MASTERS) from the
+    /// sentinels (we'll return the result of the first sentinel who responds to that
```

Comment:
remove the part in parenthesis, it's an implementation detail.

Reply:
Ok, done!